src/binaryen-c.cpp

```diff
@@ -3578,6 +3598,82 @@ void BinaryenSetMemory(BinaryenModuleRef module,
   }
 }
 
+// Memory segments
+
+uint32_t BinaryenGetNumMemorySegments(BinaryenModuleRef module) {
+  if (tracing) {
+    std::cout << "  BinaryenGetNumMemorySegments(the_module);\n";
+  }
+
+  auto* wasm = (Module*)module;
+  return wasm->memory.segments.size();
+}
+int64_t BinaryenGetMemorySegmentByteOffset(BinaryenModuleRef module,
+                                           BinaryenIndex id) {
+  if (tracing) {
+    std::cout << "  BinaryenGetMemorySegmentByteOffset(the_module, " << id
+              << ");\n";
+  }
+
+  auto* wasm = (Module*)module;
+  if (wasm->memory.segments.size() <= id) {
+    Fatal() << "invalid segment id.";
+  }
+  Memory::Segment segment = wasm->memory.segments[id];
```

Comment:
```suggestion
  Memory::Segment& segment = wasm->memory.segments[id];
```
This uses a reference, which avoids copying the segment and all its data. This happens below as well.